### PR TITLE
Hughsie/shim hive

### PR DESCRIPTION
**Pull request by hughsie**
Type of pull request:

- [ ] New plugin (Please include [new plugin checklist](https://github.com/fwupd/fwupd/wiki/New-plugin-checklist))
- [ ] Code fix
- [x] Feature
- [ ] Documentation

### CODE REVIEW

File: libfwupdplugin/fu-input-stream.c [hunk 1]
```diff
@@ -364,6 +364,40 @@ fu_input_stream_read_bytes(GInputStream *stream, gsize offset, gsize count, GErr
...
+ **/
+gchar *
+fu_input_stream_read_string(GInputStream *stream, gsize offset, gsize count, GError **error)
+{
+	g_autoptr(GByteArray) buf = NULL;
+
+	g_return_val_if_fail(G_IS_INPUT_STREAM(stream), NULL);
+	g_return_val_if_fail(error == NULL || *error == NULL, NULL);
+
+	buf = fu_input_stream_read_byte_array(stream, offset, count, error);
```
Comment (nfrayer) [inline libfwupdplugin/fu-input-stream.c:388]: Is there garbage collection or is "buf" going to leak ?

File: libfwupdplugin/fu-input-stream.c [hunk 2]
```diff
@@ -364,6 +364,40 @@ fu_input_stream_read_bytes(GInputStream *stream, gsize offset, gsize count, GErr
...
+	if (buf == NULL)
+		return NULL;
+	if (!g_utf8_validate_len((const gchar *)buf->data, buf->len, NULL)) {
```
Comment (nfrayer) [inline libfwupdplugin/fu-input-stream.c:391]: Is there garbage collection or is "buf" going to leak ?

Comment (nfrayer) [reply libfwupdplugin/fu-input-stream.c:391]: I am assuming g_autoptr is autocleaning when going out of scope

Comment (hughsie) [reply libfwupdplugin/fu-input-stream.c:391]: Exactly that -- g_autoptr() is a compile-time wrapper around `__attribute__((cleanup))` -- each GLib type uses `G_DEFINE_AUTOPTR_CLEANUP_FUNC` to tell the compiler what to run when the variable goes out of scope -- in this case it'd be something like `G_DEFINE_AUTOPTR_CLEANUP_FUNC(GByteArray, g_byte_array_unref)`

**Comment by hughsie** [comment]
@vathpela does this align with what you're planning to do with shim? It's on the list of things I want to get into 2.0.0 so a 1 minute review of just that .rs file would be most welcome.

**Comment by vathpela** [comment]
> @vathpela does this align with what you're planning to do with shim? It's on the list of things I want to get into 2.0.0 so a 1 minute review of just that .rs file would be most welcome.

The struct definitions in that .rs file look right to me.

**Pull request merged**